tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/upgrade/Elasticsearch.java

```diff
@@ -0,0 +1,153 @@
+
+    public Elasticsearch(Config cfg, Network network, Network... networks)
+    {
+        this.cfg = cfg;
+
+        additionalNetworks = Stream.of(networks).map(Network::getId).collect(Collectors.toUnmodifiableSet());
+
+        elasticsearch = new GenericContainer<>(cfg.getElasticsearchImage())
+                .withEnv("xpack.security.enabled", "false")
+                .withEnv("discovery.type", "single-node")
+                .withNetworkAliases(cfg.getElasticsearchHostname())
+                .withNetwork(network)
+                .withExposedPorts(9200);
+    }
+
+    public long getIndexedDocumentCount() throws IOException
+    {
+        final Map<?, ?> countResponse = gson.fromJson(getString("/" + cfg.getIndexName() + "/_count"), Map.class);
+        return Optional
+                .ofNullable(countResponse.get("count"))
+                .filter(Number.class::isInstance)
+                .map(Number.class::cast)
+                .map(Number::longValue)
+                .orElseThrow();
+    }
+
+    public boolean isIndexCreated() throws IOException
+    {
+        try
+        {
+            return getString("/" + cfg.getIndexName() + "/_mapping").contains("cm%3Acontent");
+        } catch (FileNotFoundException e)
+        {
+            return false;
+        }
+    }
+
+    private String getString(String path) throws IOException
+    {
+        final URL url = getRequestUrl(path);
+
+        final HttpURLConnection c = (HttpURLConnection) url.openConnection();
+        c.setRequestMethod("GET");
+        c.setConnectTimeout(ES_API_TIMEOUT_MS);
+        c.setReadTimeout(ES_API_TIMEOUT_MS);
+        c.setDoOutput(true);
+
+        try (BufferedReader r = new BufferedReader(new InputStreamReader(c.getInputStream())))
+        {
+            return r.lines().collect(Collectors.joining(System.lineSeparator()));
+        }
+    }
+
+    public void start()
+    {
+        if (elasticsearch.isRunning())
+        {
+            throw new IllegalStateException("Already started");
+        }
+
+        elasticsearch.start();
+        waitForAvailability();
+
+        connectElasticSearchToAdditionalNetworks();
+        waitForAvailability();
+    }
+
+    private URL getRequestUrl(String path)
+    {
+        try
+        {
+            final URI uri = URI
+                    .create("http://" + elasticsearch.getHost() + ":" + elasticsearch.getMappedPort(9200))
+                    .resolve(path);
+            return uri.toURL();
+        } catch (MalformedURLException e)
+        {
+            throw new RuntimeException("Failed to create a valid url.", e);
+        }
+    }
+
+    private void waitForAvailability()
+    {
+        for (int i = 0; i < 3; i++)
+        {
+            waitFor("Elasticsearch Startup", ofMinutes(1), () -> {
+                try
+                {
+                    return !isIndexCreated();
+                } catch (IOException e)
+                {
+                    return false;
```

Comment:
While with "FileNotFoundException" in the inner method I can understand ignoring the exception and returning false, here I don't understand this approach. I get that after 1 minute it won't matter since a RuntimeException will be thrown from waitFor, but shouldn't exceptions other than FileNotFound actually immediately throw such a RuntimeException instead of locking up for 1 minute?

Reply:
Nice catch! The outer loop is a leftover from a previous implementation I had in this method. I've removed it. Thank you.